File: problems-extend/1701-1800/1768_merge-strings-alternately/solution.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {string}
 */
var mergeAlternately = function (word1, word2) {
    // One pointer per word: emit alternately while both words still
    // have characters, then append whichever tail remains.
    const out = [];
    let i = 0;
    let j = 0;
    while (i < word1.length && j < word2.length) {
        out.push(word1[i]);
        out.push(word2[j]);
        i++;
        j++;
    }
    out.push(word1.slice(i), word2.slice(j));
    return out.join("");
};

File: problems-extend/1701-1800/1768_merge-strings-alternately/solution.cpp
class Solution {
public:
    string mergeAlternately(string word1, string word2) {
        // One pointer per word: emit alternately while both words still
        // have characters, then append whichever tail remains.
        string out;
        out.reserve(word1.size() + word2.size());
        size_t i = 0;
        size_t j = 0;
        while (i < word1.size() && j < word2.size()) {
            out.push_back(word1[i]);
            out.push_back(word2[j]);
            i++;
            j++;
        }
        out.append(word1, i, string::npos);
        out.append(word2, j, string::npos);
        return out;
    }
};
